File: hackathon-reverse/imports-ruby/import_rel.rb
#! /usr/bin/env ruby

def traverse_file( f )
	File.open( f, 'r', & :readlines ).map{|l|
		$1 if l =~ /^import +(?:[^ ]+ +)?([^ ]+) *;/ }.compact
end

def traverse_dir( path )
	Hash[ Dir.new( path ).reject{|f| ['.','..'].include?(f)}.map{|f|
		f = path+'/'+f
		File.directory?( f ) ? traverse_dir( f ) :
			(f.end_with?('.java') ? {f => traverse_file( f )} : nil)
	}.compact.map(& :to_a).flatten(1)]
end

if $0 == __FILE__
	path = ARGV[0] || (raise "Need directory argument.")
	rel = Dir.chdir( path ){ traverse_dir( 'org' )}
	fmt = lambda{|s| s.split('.')[0..-2].join('.').gsub('/','.')}
	puts 'digraph Relation { node [shape=box]'
	rel.each{|k,a|
		puts '"%s" [shape=ellipse]' % [fmt[k]]
		a.each{|v| puts '"%s" -> "%s"' % [fmt[k], v] }
	}
	puts '}'
end

File: hackathon/imports-python/importRel.py
#! /usr/bin/env python

import os, sys
rel = {}

def traverseFile(f):
	global path
	if not rel.has_key(f): rel[f] = []
	txt = open(path+f,'r')
	for l in txt.readlines():
		if l.startswith('import '):
			rel[f].append(l.split(' ')[-1].strip()[:-1])
	txt.close()

def traverseDir(d):
	global path
	for f in os.listdir(path+d):
		if os.path.isdir(path+d+'/'+f): traverseDir(d+'/'+f)
		elif not f.endswith('.class'):  traverseFile(d+'/'+f)

if __name__ == '__main__':
	path = sys.argv[1]+'/'
	traverseDir('org')
	print 'digraph Relation { node [shape=box]'
	for k in rel.keys():
		print '"%s" [shape=ellipse]' % (k.split('.')[0].replace('/','.'))
		for v in rel[k]:
			print '"%s" -> "%s"' % (k.split('.')[0].replace('/','.'),v)
	print '}'
